### ISSUE TITLE
Get/Create Texture from AtlastTexture or Sprite Region

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot-proposals#3882]
### Describe the project you are working on

2D platform

### Describe the problem or limitation you are having in your project

Unless there is a way I couldn't find, it should be possible/would be nice in Godot to treat regions of Sprites, or AtlastTextures, as Textures. Or at least there be a simple means of creating a Texture from a region of a sprite/AtlastTexture.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Typically you use sprite sheets to represent animations or your entire game graphics and use them in game by directly specifying regions, using AnimatedSprites or using .tres files to represent single items. But doing so means you cannot use these .tres/regions where functions require actual Textures.

For example, pretend you tried to do this manually in code by converting a Sprite's texture/region to an AtlasTexture:

```
shard_texture=AtlasTexture.new()
shard_texture.atlas=my_sprite.texture
shard_texture.region=my_sprite.region_rect
```

You can easily extract the whole texture using get_atlas(), but you're still left without any ability to extract a Texture representing the region from this as AtlastTexture<>Texture and functions requiring Texture will fail, e.g. you've created a Polygon2D or Sprite that you want to set a texture to, but get_atlas() would end up showing the entire sheet not the region.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Sprite.get_texture_from_region()
AtlasTexture.get_texture()

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not that I know of. What I have to do now is duplicate everything I have in my sprite sheets as individual graphic files that I use in code where a Texture is required.

### Is there a reason why this should be core and not an add-on in the asset library?

Sprite sheets are great, but they're not much use when you cannot get a specific region as a texture.

**Comment by KoBeWi** [comment]
You can do this by creating Image, copying part of the texture original and then creating ImageTexture.

**Comment by chucklepie** [comment]
@KoBeWi So in order to get a texture region from a Sprite or some other object with a texture I have to get the texture of the object, create an Image, Create an ImageTexture, cut the image and apply it. Has somebody lost sight of something here, everything in Godot is centred on textures but there's no simple means of getting them from things that are based on them ;-)

I mean, AtlasTexture not only has the word Texture in the name but it serves solely to create a region of a texture, but you can't even get the texture from it....

**Comment by KoBeWi** [comment]
But why do you even want to get the region? You can use AtlasTexture in most nodes (Sprite, TextureRect etc.) as texture without problems.

**Comment by chucklepie** [comment]
Ok, so presuming this is what you mean, it doesn't work, it might be an AtlastTexture but shard_texture still ends up with the entire texture of the atlas not the region you specify.

```
		var a=AtlasTexture.new()
		a.atlas=shard_node.texture
		a.region=shard_node.region_rect
		var shard_texture:Texture=a
```

**Comment by KoBeWi** [comment]
But when you use `shard_texture` in e.g. TextureRect, it will only display the region. So what is the problem?

**Comment by jordigcs** [comment]
> Ok, so presuming this is what you mean. Presume 'shard_node' is a sprite with a region set on a larger sprite, e.g. it's a .tres file or manually set region of a larger sheet.
> 
> ```
>                var shard_texture:Texture=null
> 		var a=AtlasTexture.new()
> 		a.atlas=shard_node.texture
> 		a.region=shard_node.region_rect
> 		shard_texture=a
> ```
> 
> It doesn't work, it might be an AtlastTexture but shard_texture still ends up with the entire texture of the atlas not the region the atlas specifies.

Here's the correct code to do this:
```swift
var img_tex = ImageTexture.new();
img_tex.create_from_image(texture.atlas.get_image().get_rect(texture.region));
```

**Comment by chucklepie** [comment]
Ok, my fault then. Not exactly an obvious thing though :)

**Comment by KoBeWi** [comment]
Yes. AtlasTexture won't work with e.g. AnimatedTexture or Particles2D, but ImageTexture can be used anywhere.

**Comment by jordigcs** [comment]
> Ok, my fault then. Not exactly an obvious thing though :)
> 
> So that code will create a normal texture that can be set, e.g. to a polygon2D texture?

No problem, I actually think it would be a good convinience feature, and actually even implemented it! I was about to make a PR for it before I checked the blog and found that the PR would most likely be denied for the reason I listed.

**Comment by stanleyHorace** [comment]
Although old, I figure that since this is still relevant I'll crosslink to the issue with more data on it:

https://github.com/godotengine/godot/issues/43268#issuecomment-720571636

AtlasTextures in Godot have some known problems, I'll quote [the relevant comment](https://github.com/godotengine/godot/issues/43268#issuecomment-720571636) from 2020:

> 
> Polygon2D (
> https://github.com/godotengine/godot/issues/25049)
> MeshInstance, MultiMeshInstance and CSG nodes (
> https://github.com/godotengine/godot-proposals/issues/1708)
> Shaders (
> https://github.com/godotengine/godot/issues/25895)
> NinePatchRect (partially fixed by
> https://github.com/godotengine/godot/pull/21156 because it's slightly simpler to implement there, but some configurations can still break it)
> Repeat/Clamp won't work in general
> https://github.com/godotengine/godot/issues/5660, which was flagged as documentation issue
> And if we are at it, get_data() will also return the whole texture, and get_rid() will give you the RID of the whole texture too.

**Comment by KoBeWi** [comment]
> Here's the correct code to do this:

That code won't work anymore. Here's an updated one:
```GDScript
image_texture = ImageTexture.create_from_image(atlas_texture.get_image())
```
